### PR TITLE
 added extra encryption algorithm

### PR DESCRIPTION
**Pull request by samalloing**
Hi @carlwilson 

In this pull request I added code for a new encryption algorithm 256-bit AES. It is allowed in the extensionlevel 3 as a new encryption algorithm. This code is independent of the Extension level code I mailed you about.

Sam

### CODE REVIEW

File: jhove-modules/pdf-hul/src/main/java/edu/harvard/hul/ois/jhove/module/PdfModule.java [hunk 1]
```diff
@@ -1863,6 +1865,37 @@ protected boolean readEncryptDict(RepInfo info) throws IOException {
...
+									toHex(((PdfSimpleObject) oeObj).getRawBytes())));
+						}
+					} else {
+						// if algValue is 5; OE is mandatory
+						throw new PdfInvalidException // PDF-HUL-15X
+						(MessageConstants.PDF_HUL_152, _parser.getOffset());
+					}
+				}
+				// Required if ExtensionLevel 3 and Encryption Algorithm (V) is 5
+				if (algValue == 5) {
```
Comment (david-russo) [inline jhove-modules/pdf-hul/src/main/java/edu/harvard/hul/ois/jhove/module/PdfModule.java]: These three lines could be removed to make the following code a part of the preceding and identical `if` condition, only requiring the check to happen once.

File: jhove-modules/pdf-hul/src/main/java/edu/harvard/hul/ois/jhove/module/PdfModule.java [hunk 2]
```diff
@@ -1863,6 +1865,37 @@ protected boolean readEncryptDict(RepInfo info) throws IOException {
...
+					PdfObject oeObj = dict.get("OE");
+					if (oeObj != null) {
+						if (oeObj instanceof PdfSimpleObject) {
+							stdList.add(new Property(PROP_NAME_OWNERKEY_STRING,
+									PropertyType.STRING,
+									toHex(((PdfSimpleObject) oeObj).getRawBytes())));
+						}
+					} else {
+						// if algValue is 5; OE is mandatory
+						throw new PdfInvalidException // PDF-HUL-15X
```
Comment (david-russo) [inline jhove-modules/pdf-hul/src/main/java/edu/harvard/hul/ois/jhove/module/PdfModule.java]: These message ID markers (`// PDF-HUL-15X`) are a hold over from a refactoring that took place some time ago and are no longer necessary now that message IDs are encoded in `MessageConstants` names (like below). Feel free to remove them from any code you happen to work on, or leave them — no real issue here.

File: jhove-modules/pdf-hul/src/main/resources/edu/harvard/hul/ois/jhove/module/pdf/ErrorMessages.properties [hunk 3]
```diff
@@ -153,3 +153,6 @@ PDF-HUL-147 = Page tree node not found.
 PDF-HUL-148 = PDF minor version number is greater than 7.
 PDF-HUL-149 = Invalid indirect destination - referenced object ''{0}'' cannot be found
 PDF-HUL-150 = Cross-reference stream must be a stream
+PDF_HUL_151 = Unexpected error occurred while attempting to read the cross-reference table
+PDF_HUL_152 = The mandatory OE is missing
+PDF_HUL_153 = The mandatory UE is missing
```
Comment (david-russo) [inline jhove-modules/pdf-hul/src/main/resources/edu/harvard/hul/ois/jhove/module/pdf/ErrorMessages.properties]: To add a little more detail and bring them in line with other similar PDF messages:
"Encrypt dictionary has no OE key or it has a null value"
"Encrypt dictionary has no UE key or it has a null value"

**Review by david-russo — COMMENTED** [review]
More good work. Thanks, Sam.

**Comment by samalloing** [comment]
Thanks again @david-russo . I updated the pull request

**Review by david-russo — APPROVED** [review]

**Comment by carlwilson** [comment]
Thanks both @samalloing and @david-russo

**Comment by carlwilson** [comment]
Interesting, this has [broken the build](https://travis-ci.org/github/openpreserve/jhove/builds/697042154). The issue is that the IDs for the message resources are mal-formed and the messages can't be loaded. I've added a review comment that requests changes.

File: jhove-modules/pdf-hul/src/main/resources/edu/harvard/hul/ois/jhove/module/pdf/ErrorMessages.properties [hunk 4]
```diff
@@ -153,3 +153,6 @@ PDF-HUL-147 = Page tree node not found.
 PDF-HUL-148 = PDF minor version number is greater than 7.
 PDF-HUL-149 = Invalid indirect destination - referenced object ''{0}'' cannot be found
 PDF-HUL-150 = Cross-reference stream must be a stream
+PDF_HUL_151 = Unexpected error occurred while attempting to read the cross-reference table
```
Comment (carlwilson) [inline jhove-modules/pdf-hul/src/main/resources/edu/harvard/hul/ois/jhove/module/pdf/ErrorMessages.properties]: These are the culprits, they should be using hyphens, not underscores.

**Review by carlwilson — CHANGES_REQUESTED** [review]
See comment inline but good work overall thanks @samalloing

**Comment by samalloing** [comment]
I updated it, sorry about the mistake @carlwilson

**Comment by carlwilson** [comment]
> I updated it, sorry about the mistake @carlwilson

No need at all, that's what the autotests are for, thanks @samalloing

**Review by carlwilson — APPROVED** [review]

**Pull request merged**